### DISCUSSION TITLE
Weekly Report 09.05 - 16.05.

### DISCUSSION THREAD

**Discussion by Umbranecis** [SE-TINF22B6/Plapy#135] [Weekly reports]
Greetings,

This week we set up our testing framework so we can implement meaningful tests for out Bot.
We also continued refactoring the code to make it more accessible.

About our testing framework:

The heart of our project is the discord.js framework. It is an open source package that everyone can access and use freely. However, it is neither coded by us, nor do we have a chance to change anything about it. Because of that, testing the framework itself is very difficult and also not really necessary. Because of this, we decided to only test the interfaces accessing the framework...
For this, we had to set up mock data simulating correct discord messages and interactions. We use jest for this. Our whole Database and Discord server interactions get mocked as jest functions and return dummy values. These dummy values can be tested with unit tests. 

So far, we created unit tests for:

The SavedPlaylist Structure:
- Add songs to a savedPlaylist
- save a savedPlaylist to the database
- iterate through the songs ion a savedPlaylist

The MusicQueue Structure
- Add songs to the Music Queue
- create a queue from an interaction
- add multiple songs to the music queue at once

Unit tests to be implemented:
MusicQueue
- get created without an interaction without any default value present...

For the last, we have not yet found any way to do this. We found some workarounds, but we want to implement this further.

We also set up some Integration Tests to test if the API works correctly. To do so, we test the methods called by the API.
Here, we once again have to mock all discord and Database Elements and can only test the logic that we manually wrote, but we have no way to test the interactions with discord.js.

Since discord.js is a pretty huge framework that gets updated and maintained a lot, we can trust in it working correctly. We can only use its functionality.

In addition to tests, we also refactored the SavedPlaylist Structure this week.

First of all, playlists now have two identifying keys, one is the name, the other is the serverId (guildId in discord). Because of that, playlists now are server specific.

Additional, we changed its interaction with the addToPlaylist command:
Before this week, the SavedPlaylist had a constructor that created it with an empty song array. Thus, in the addToPlaylist command, we had to check if a playlist already exists with the given name and if so, we had to fetch it and add the song. otherwise, we had to create a new one and add the song to it. This is a lot of functionality that is in the command and leaves room for error.

The new constructor in the savedPlaylist now is private and there is a "getOrCreatePlaylist" method, that with a given name and guildID, that always returns a playlist, either a newly created one, or an existing one if one is present.
The "addSong" method now syncs the playlist to the database. Because of this, we can now only can "getOrCreatePlaylist().addSong()" in the command and the song gets added to a new or existing playlist.

Were happy to hear your feedback and were going to provide a graphical image to show our testing progress until next week

See you in class,

Team Plapy

**Comment by denniskp** [comment]
Hello Team Plapy,
your testing plan looks solid, especially considering the challenges with discord.js. It's great to see your approach in identifying potential limitations and planning workarounds. Simplifying the SavedPlaylist structure was also a smart move. Excited to see your progress next week! 👍
\- Team DHBWhub

**Comment by Ossama1203** [comment]
Hello Team Plapy,

I think you've done a good work and good advancement on "The SavedPlaylist Structure" and "The MusicQueue Structure". Do you gonna add function about the sharing music between users ?

Good luck mates
